File: moedict-android/app/src/main/java/com/yichiuan/moedict/ui/RepositoryViewModelFactory.java
package com.yichiuan.moedict.ui;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;

import com.yichiuan.moedict.data.MoeRepository;
import com.yichiuan.moedict.ui.main.MoeViewModel;
import com.yichiuan.moedict.ui.search.SearchViewModel;

public class RepositoryViewModelFactory implements ViewModelProvider.Factory {

    private MoeRepository moeRepository;

    public RepositoryViewModelFactory(MoeRepository moeRepository) {
        this.moeRepository = moeRepository;
    }

    @Override
    public <T extends ViewModel> T create(Class<T> modelClass) {
        if (modelClass.isAssignableFrom(SearchViewModel.class)) {
            return (T) new SearchViewModel(moeRepository);
        } else if (modelClass.isAssignableFrom(MoeViewModel.class)) {
            return (T) new MoeViewModel(moeRepository);
        }
        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: moedict-android/app/src/main/java/com/yichiuan/moedict/ui/RepositoryViewModelFactory.kt
package com.yichiuan.moedict.ui

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider

import com.yichiuan.moedict.data.MoeRepository
import com.yichiuan.moedict.ui.main.MoeViewModel
import com.yichiuan.moedict.ui.search.SearchViewModel

class RepositoryViewModelFactory(private val moeRepository: MoeRepository) : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(SearchViewModel::class.java)) {
            return SearchViewModel(moeRepository) as T
        } else if (modelClass.isAssignableFrom(MoeViewModel::class.java)) {
            return MoeViewModel(moeRepository) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}
